File: Tree/AVLTree/AVLTree.cs
using System;

public class AVLNode
{
    public int Key { get; set; }
    public AVLNode Left { get; set; }
    public AVLNode Right { get; set; }
    public int Height { get; set; }

    public AVLNode(int key)
    {
        Key = key;
        Left = null;
        Right = null;
        Height = 1;
    }
}

public class AVLTree
{
    private AVLNode root;

    private int Height(AVLNode node)
    {
        if (node == null)
            return 0;
        return node.Height;
    }

    private int UpdateHeight(AVLNode node)
    {
        if (node == null)
            return 0;
        node.Height = 1 + Math.Max(Height(node.Left), Height(node.Right));
        return node.Height;
    }

    private int BalanceFactor(AVLNode node)
    {
        if (node == null)
            return 0;
        return Height(node.Left) - Height(node.Right);
    }

    private AVLNode RightRotate(AVLNode y)
    {
        AVLNode x = y.Left;
        AVLNode T2 = x.Right;

        x.Right = y;
        y.Left = T2;

        UpdateHeight(y);
        UpdateHeight(x);

        return x;
    }

    private AVLNode LeftRotate(AVLNode x)
    {
        AVLNode y = x.Right;
        AVLNode T2 = y.Left;

        y.Left = x;
        x.Right = T2;

        UpdateHeight(x);
        UpdateHeight(y);

        return y;
    }

    private AVLNode Balance(AVLNode node, int key)
    {
        UpdateHeight(node);
        int balance = BalanceFactor(node);

        // Left Heavy
        if (balance > 1 && key < node.Left.Key)
            return RightRotate(node);

        // Right Heavy
        if (balance < -1 && key > node.Right.Key)
            return LeftRotate(node);

        // Left-Right Heavy
        if (balance > 1 && key > node.Left.Key)
        {
            node.Left = LeftRotate(node.Left);
            return RightRotate(node);
        }

        // Right-Left Heavy
        if (balance < -1 && key < node.Right.Key)
        {
            node.Right = RightRotate(node.Right);
            return LeftRotate(node);
        }

        return node;
    }

    private AVLNode Insert(AVLNode node, int key)
    {
        if (node == null)
            return new AVLNode(key);

        if (key < node.Key)
            node.Left = Insert(node.Left, key);
        else if (key > node.Key)
            node.Right = Insert(node.Right, key);
        else
            return node;  // Duplicate keys are not allowed

        return Balance(node, key);
    }

    private void InOrderTraversal(AVLNode node)
    {
        if (node != null)
        {
            InOrderTraversal(node.Left);
            Console.Write(node.Key + " ");
            InOrderTraversal(node.Right);
        }
    }

    public void Insert(int key)
    {
        root = Insert(root, key);
    }

    public void PrintInOrderTraversal()
    {
        InOrderTraversal(root);
    }
}

class Program
{
    static void Main()
    {
        AVLTree avlTree = new AVLTree();
        avlTree.Insert(10);
        avlTree.Insert(5);
        avlTree.Insert(15);
        avlTree.Insert(3);
        avlTree.Insert(7);

        Console.Write("Inorder traversal: ");
        avlTree.PrintInOrderTraversal();
        Console.WriteLine();
    }
}

File: Tree/AVLTree/AVLTree.cpp
#include <iostream>

class AVLNode {
public:
    int key;
    AVLNode* left;
    AVLNode* right;
    int height;

    AVLNode(int k) : key(k), left(nullptr), right(nullptr), height(1) {}
};

class AVLTree {
private:
    AVLNode* root;

    int height(AVLNode* node) {
        if (!node)
            return 0;
        return node->height;
    }

    int updateHeight(AVLNode* node) {
        if (!node)
            return 0;
        node->height = 1 + std::max(height(node->left), height(node->right));
        return node->height;
    }

    int balanceFactor(AVLNode* node) {
        if (!node)
            return 0;
        return height(node->left) - height(node->right);
    }

    AVLNode* rightRotate(AVLNode* y) {
        AVLNode* x = y->left;
        AVLNode* T2 = x->right;

        x->right = y;
        y->left = T2;

        updateHeight(y);
        updateHeight(x);

        return x;
    }

    AVLNode* leftRotate(AVLNode* x) {
        AVLNode* y = x->right;
        AVLNode* T2 = y->left;

        y->left = x;
        x->right = T2;

        updateHeight(x);
        updateHeight(y);

        return y;
    }

    AVLNode* balance(AVLNode* node, int key) {
        updateHeight(node);
        int balance = balanceFactor(node);

        // Left Heavy
        if (balance > 1 && key < node->left->key)
            return rightRotate(node);

        // Right Heavy
        if (balance < -1 && key > node->right->key)
            return leftRotate(node);

        // Left-Right Heavy
        if (balance > 1 && key > node->left->key) {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }

        // Right-Left Heavy
        if (balance < -1 && key < node->right->key) {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }

        return node;
    }

    AVLNode* insert(AVLNode* node, int key) {
        if (!node)
            return new AVLNode(key);

        if (key < node->key)
            node->left = insert(node->left, key);
        else if (key > node->key)
            node->right = insert(node->right, key);
        else
            return node;  // Duplicate keys are not allowed

        return balance(node, key);
    }

    void inorderTraversal(AVLNode* node) {
        if (node) {
            inorderTraversal(node->left);
            std::cout << node->key << " ";
            inorderTraversal(node->right);
        }
    }

public:
    AVLTree() : root(nullptr) {}

    void insert(int key) {
        root = insert(root, key);
    }

    void printInOrderTraversal() {
        inorderTraversal(root);
    }
};

int main() {
    AVLTree avlTree;
    avlTree.insert(10);
    avlTree.insert(5);
    avlTree.insert(15);
    avlTree.insert(3);
    avlTree.insert(7);

    std::cout << "Inorder traversal: ";
    avlTree.printInOrderTraversal();
    std::cout << std::endl;

    return 0;
}